### PR TITLE
Workspace ap4

### PR DESCRIPTION
**Pull request by haogatyp**
Contains the tickets CMR-703, CMR-685 and CMR-694.

### CODE REVIEW

File: Classes/Services/ElasticSearch/ElasticSearch.php [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+     *
+     * @var \EWW\Dpf\Helper\FrontendUserHelper
+     * @inject
+     */
+    protected $frontendUserHelper = null;
+
+
+    protected $client;
+
+    protected $server = 'host.docker.internal'; //127.0.0.1';
```
Comment (claussni) [inline Classes/Services/ElasticSearch/ElasticSearch.php:38]: [code] Do we still have a plugin configuration section for the Elasticsearch connection parameters?

File: Classes/Services/ElasticSearch/ElasticSearch.php [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+            $params = [
+                'refresh' => 'wait_for',
+                'index' => $this->indexName,
+                'id' => $identifier
+            ];
+
+            $this->client->delete($params);
+
+        } catch (\Exception $e) {
+            // Nothing to be done.
```
Comment (claussni) [inline Classes/Services/ElasticSearch/ElasticSearch.php:243]: [code] This is the perfect opportunity for logging.

File: Configuration/TypoScript/constants.txt [hunk 3]
```diff
@@ -50,5 +50,11 @@ plugin.tx_dpf {
 
 			# cat=plugin.tx_dpf/link; type=int+; label=PID of the landing page (Backend Preview)
 			previewPid =
+
+		    # cat=plugin.tx_dpf/link; type=int+; label=Workspace items per page (Backoffice)
+            workspaceItemsPerPage = 30
+
+		    # cat=plugin.tx_dpf/link; type=int+; label=University collection value (Backoffice)
+            universityCollection = epflicht
```
Comment (claussni) [inline Configuration/TypoScript/constants.txt:58]: What is the function of "universityCollection"?

File: Build/docker-compose.yml [hunk 4]
```diff
@@ -35,12 +35,35 @@ services:
...
+            - elasticsearch-kitodo-75:/elasticsearch/data
+        ulimits:
+            memlock:
+                soft: -1
+                hard: -1
+        ports:
+            - "4713:9200"
+        networks:
+            - backend
+            - kitodopublication_frontend
```
Comment (claussni) [inline Build/docker-compose.yml:58]: [code] Wrong networt name. The defined network from the 'networks' section is 'frontend'.

File: Classes/Domain/Workflow/DocumentWorkflow.php [hunk 5]
```diff
@@ -22,57 +22,107 @@
...
+        DocumentWorkflow::SIMPLE_STATE_IN_PROGRESS => [
+            DocumentWorkflow::STATE_IN_PROGRESS_NONE,
+            DocumentWorkflow::STATE_IN_PROGRESS_ACTIVE,
+            DocumentWorkflow::STATE_IN_PROGRESS_INACTIVE,
+            DocumentWorkflow::STATE_IN_PROGRESS_DELETED
+        ],
+        DocumentWorkflow::SIMPLE_STATE_RELEASED => [DocumentWorkflow::STATE_NONE_ACTIVE]
+    ];
+
+    const PLACES = [
```
[the remark this answers is not in the document]
Comment (haogatyp) [reply Classes/Domain/Workflow/DocumentWorkflow.php:125]: To implement the workflow and its states the symfony component "workflow" was used. The wording "places" has been taken from the terminologie of that component. Places are the states a document can have. A single state consists of the local and the connected state.

File: Classes/Services/ElasticSearch/ElasticSearch.php (hunk 1 above)
Comment (haogatyp) [reply Classes/Services/ElasticSearch/ElasticSearch.php:38]: Yes, it still has to be configured within the client dataset in the list module.

File: Configuration/TypoScript/constants.txt (hunk 3 above)
Comment (haogatyp) [reply Configuration/TypoScript/constants.txt:58]: If a publiakation belongs to the configuered unviversity collection it has been emerged from scientific work at the university. In our case this is the "UL". See also Ticket CMR-685.

File: Build/docker-compose.yml [hunk 6]
```diff
@@ -35,12 +35,35 @@ services:
             - frontend
             - backend
 
+    elasticsearch-kitodo-75:
+        image: docker.elastic.co/elasticsearch/elasticsearch:7.5.0
+        container_name: elasticsearch-kitodo-75
+        environment:
+            - xpack.ml.enabled: false
```
[the remark this answers is not in the document]
Comment (haogatyp) [reply Build/docker-compose.yml]: Fixed.

**Comment by claussni** [comment]
> Docker Compose fails when building the images:
> 
> ```
> Step 6/11 : FROM typo3builder
>  ---> 46d694515740
> Step 7/11 : WORKDIR /var/www/html
>  ---> Running in 6c20332587b1
> Removing intermediate container 6c20332587b1
>  ---> 11e64d435183
> Step 8/11 : RUN composer config repositories.kitodo-publication '{"type": "path", "url": "/app", "options": {"symlink": true}}' &&     composer config minimum-stability dev &&     composer config prefer-stable true
>  ---> Running in 8f1f25d5cb9d
> Removing intermediate container 8f1f25d5cb9d
>  ---> 582a38a8ba0e
> Step 9/11 : RUN composer require devlog/devlog:~3.0.4 &&     composer require kitodo/presentation
>  ---> Running in 49575b2e1164
> 
>                                                                     
>   [RuntimeException]                                                
>   The `url` supplied for the path (/app) repository does not exist  
>                                                                     
> 
> require [--dev] [--prefer-source] [--prefer-dist] [--fixed] [--no-progress] [--no-suggest] [--no-update] [--no-scripts] [--update-no-dev] [--update-with-dependencies] [--update-with-all-dependencies] [--ignore-platform-reqs] [--prefer-stable] [--prefer-lowest] [--sort-packages] [-o|--optimize-autoloader] [-a|--classmap-authoritative] [--apcu-autoloader] [--] [<packages>]...
> 
> ERROR: Service 'typo3' failed to build: The command '/bin/sh -c composer require devlog/devlog:~3.0.4 &&     composer require kitodo/presentation' returned a non-zero code: 1
> ```

This seems to be the problem: #167

**Comment by claussni** [comment]
Elasticsearch container is not started after `docker-compose up`. Errors from the containe log output:
```
{"type": "server", "timestamp": "2020-05-15T06:09:39,225Z", "level": "INFO", "component": "o.e.b.BootstrapChecks", "cluster.name": "es-docker-cluster", "node.name": "elasticsearch-kitodo-75", "message": "bound or publishing to a non-loopback address, enforcing bootstrap checks" }
ERROR: [1] bootstrap checks failed
[1]: max virtual memory areas vm.max_map_count [65530] is too low, increase to at least [262144]
```

File: Build/docker-compose.yml (hunk 4 above)
Comment (haogatyp) [reply Build/docker-compose.yml]: Network should be correct now.

**Comment by haogatyp** [comment]
> vm.max_map_count

I assume this can only be configured in the host and not inside a docker container.

**Comment by claussni** [comment]
Elasticsearch has Docker prerequisites: https://www.elastic.co/guide/en/elasticsearch/reference/master/docker.html#docker-prod-prerequisites

**Comment by claussni** [comment]
In principle docker allows to set Kernel parameters on container start. This is limited to certain namespaces. The parameter in question here (vm.max_map_count) is not in this namespaces because it is a host parameter.

**Comment by claussni** [comment]
I thought there is no Backend Module for Kitodo.Publication anymore. Why this:
Screenshot from 2020-05-18 11-16-22?

File: Classes/Services/ElasticSearch/ElasticSearch.php (hunk 2 above)
Comment (haogatyp) [reply Classes/Services/ElasticSearch/ElasticSearch.php]: Logging has been added.

**Comment by haogatyp** [comment]
> I thought there is no Backend Module for Kitodo.Publication anymore. Why this:
> Screenshot from 2020-05-18 11-16-22?



> I thought there is no Backend Module for Kitodo.Publication anymore. Why this:
> Screenshot from 2020-05-18 11-16-22?

The backend module no longer exists, thats right, but there was still a definition for the entry in the backend module column. This has now been removed and should not appear anymore.

File: Classes/Helper/DocumentValidator.php [hunk 7]
```diff
@@ -70,20 +78,33 @@ protected function hasFieldWithValue(DocumentFormGroup $group)
     /**
      * @param DocumentFormGroup $group
      * @param bool $hasFiles
+     * @param bool $strict : If false, invisible form fields are not validated.
```
[the remark this answers is not in the document]
Comment (haogatyp) [reply Classes/Helper/DocumentValidator.php]: Has been renamed into $validateInvisableFields

File: Classes/Helper/FrontendUserHelper.php [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+ *
+ * For the full copyright and license information, please read the
+ * LICENSE.txt file that was distributed with this source code.
+ *
+ * The TYPO3 project - inspiring people to share!
+ */
+
+use EWW\Dpf\Security\Security;
+
+class FrontendUserHelper
```
[the remark this answers is not in the document]
Comment (haogatyp) [reply Classes/Helper/FrontendUserHelper.php]: The functionality has been moved into the FrontendUser class.

File: Classes/Domain/Workflow/DocumentWorkflow.php [hunk 9]
```diff
@@ -22,57 +22,107 @@
...
+    const TRANSITION_POSTPONE            = "POSTPONE_TRANSITION";
+    const TRANSITION_RELEASE_PUBLISH     = "RELEASE_PUBLISH_TRANSITION";
+    const TRANSITION_RELEASE_ACTIVATE    = "RELEASE_ACTIVATE_TRANSITION";
+    const TRANSITION_REMOTE_UPDATE       = "REMOTE_UPDATE_TRANSITION";
+    const TRANSITION_IN_PROGRESS         = "IN_PROGRESS_TRANSITION";
+    const TRANSITION_DELETE_LOCALLY      = "DELETE_LOCALLY_TRANSITION";
+    const TRANSITION_DELETE_WORKING_COPY = "DELETE_WORKING_COPY_TRANSITION";
+    const TRANSITION_DELETE_DISCARDED    = "DELETE_DISCARDED_TRANSITION";
+
+    const SIMPLE_STATE_NEW = "new";
```
[the remark this answers is not in the document]
Comment (haogatyp) [reply Classes/Domain/Workflow/DocumentWorkflow.php]: Simple state was renamed into alias state. This also affects the elasticsearch index, therefore the index "kitodo_publication" has to be rebuild.

File: Resources/Private/Language/locallang.xlf [hunk 10]
```diff
@@ -546,17 +567,152 @@
...
+            <source>User</source>
+        </trans-unit>
+        <trans-unit id="manager.workspace.creator.unknown">
+            <source>Unknown</source>
+        </trans-unit>
+        <trans-unit id="manager.workspace.creator.tooltip.self">
+            <source>Record created by myself.</source>
+        </trans-unit>
+        <trans-unit id="manager.workspace.creator.tooltip.librarian">
+            <source>Von der Bibliothek erstellter Datensatz.</source>
```
[the remark this answers is not in the document]
Comment (haogatyp) [reply Resources/Private/Language/locallang.xlf]: Target and source have been corrected.

File: Resources/Private/Language/de.locallang_db.xlf [hunk 11]
```diff
@@ -569,6 +565,30 @@
...
+            <source>Document Identifier</source>
+            <target>Dokument-Bezeichner</target>
+        </trans-unit>
+        <trans-unit id="tx_dpf_domain_model_editinglock.editor_uid" approved="yes">
+            <source>Editor UID</source>
+            <target>Bearbeiter-UID</target>
+        </trans-unit>
+        <trans-unit id="tx_dpf_domain_model_bookmark" approved="yes">
+            <source>Bookmark</source>
+            <target>Bookmark</target>
```
[the remark this answers is not in the document]
Comment (haogatyp) [reply Resources/Private/Language/de.locallang_db.xlf:582]: I guess this is on purpose, we currently use the term "Bookmark" in the German user interface.

File: Classes/Controller/DocumentFormBackofficeController.php [hunk 12]
```diff
@@ -139,7 +171,7 @@ public function createSuggestionDocumentAction(\EWW\Dpf\Domain\Model\DocumentFor
 
         if ($workingCopy->isTemporary()) {
             $workingCopy->setTemporary(false);
-            $workingCopy->setEditorUid(0);
+            // +++delete+++ $workingCopy->setEditorUid(0);
```
[the remark this answers is not in the document]
Comment (haogatyp) [reply Classes/Controller/DocumentFormBackofficeController.php]: The comment has been remove from the code.

File: Classes/Security/DocumentVoter.php [hunk 13]
```diff
@@ -223,7 +226,7 @@ public function voteOnAttribute($attribute, $subject = NULL)
                 break;
         }
 
-        throw new \Exception('An unexpected error occurred!');
+        throw new \Exception("This code should not be reached!");
```
[the remark this answers is not in the document]
Comment (haogatyp) [reply Classes/Security/DocumentVoter.php:229]: The function now returns false instead, which leads to an flash error message in the frontend. In addition this unexpected event will now be logged.

**Pull request merged**